src/core/lib/iomgr/tcp_posix.cc

```diff
@@ -79,6 +90,178 @@ typedef size_t msg_iovlen_type;
 
 extern grpc_core::TraceFlag grpc_tcp_trace;
 
+namespace grpc_core {
+
+class TcpZcpSendRecord {
+ public:
+  struct OutgoingOffset {
+    size_t slice_idx = 0;
+    size_t byte_idx = 0;
+  };
+
+  TcpZcpSendRecord() { grpc_slice_buffer_init(&buf_); }
+  ~TcpZcpSendRecord() {
+    AssertEmpty();
+    grpc_slice_buffer_destroy_internal(&buf_);
+  }
+  OutgoingOffset& GetOutgoingOffsets() { return out_offset_; }
+  grpc_slice_buffer* Buffer() { return &buf_; }
+  void PrepareForSends(grpc_slice_buffer* slices_to_send) {
+    AssertEmpty();
+    out_offset_.slice_idx = 0;
+    out_offset_.byte_idx = 0;
+    grpc_slice_buffer_swap(slices_to_send, &buf_);
+    Ref();
+  }
+  intptr_t dbg() { return ref_.Load(MemoryOrder::ACQUIRE); }
+  void Ref() { ref_.FetchAdd(1, MemoryOrder::RELAXED); }
+  bool Unref() {
+    const intptr_t prior = ref_.FetchSub(1, MemoryOrder::ACQ_REL);
+    GPR_DEBUG_ASSERT(prior > 0);
+    if (prior == 1) {
+      AllSendsComplete();
+      return true;
+    }
+    return false;
+  }
+
+ private:
+  void AssertEmpty() {
+    GPR_DEBUG_ASSERT(buf_.count == 0);
+    GPR_DEBUG_ASSERT(buf_.length == 0);
+    GPR_DEBUG_ASSERT(ref_.Load(MemoryOrder::RELAXED) == 0);
+  }
+  void AllSendsComplete() {
+    GPR_DEBUG_ASSERT(ref_.Load(MemoryOrder::RELAXED) == 0);
+    grpc_slice_buffer_reset_and_unref_internal(&buf_);
+  }
+
+  grpc_slice_buffer buf_;
+  Atomic<intptr_t> ref_;
+  OutgoingOffset out_offset_;
+};
+
+class TcpZcpSendCtx {
+ public:
+  class LockGuard {
+   public:
+    explicit LockGuard(gpr_spinlock* lock) : lock_(lock) {
+      gpr_spinlock_lock(lock_);
+    }
+    ~LockGuard() { gpr_spinlock_unlock(lock_); }
+
+    LockGuard(const LockGuard&) = delete;
+    LockGuard& operator=(const LockGuard&) = delete;
+
+   private:
+    gpr_spinlock* const lock_;
+  };
+
+  static constexpr int kMaxSends = 4;
```

Comment:
Is there any reason why the max number of sends is 4?

Reply:
To limit the maximum amount of memory pinned. In a pathological case, suppose the kernel is really slow getting back with callbacks; if we didn't limit the max sends, we could OOM.

Reply:
I did make this configurable in the latest version.